accepted/2021/System.Diagnostics/Metrics-Design.md

```diff
@@ -0,0 +1,484 @@
+                            string name,
+                            Func<T> observeValue,
+                            string? description = null,
+                            string? unit = null) where T : unmanaged { throw null; }
+
+        public ObservableGauge<T> CreateObservableGauge<T>(
+                            string name,
+                            Func<Measurement<T>> observeValue,
+                            string? description = null,
+                            string? unit = null) where T : unmanaged { throw null; }
+
+        public ObservableGauge<T> CreateObservableGauge<T>(
+                            string name,
+                            Func<IEnumerable<Measurement<T>>> observeValues,
+                            string? description = null,
+                            string? unit = null) where T : unmanaged { throw null; }
+
+        /// <summary>
+        /// Factory Methods to create observable UpDownCounter instrument.
+        /// </summary>
+        public ObservableUpDownCounter<T> CreateObservableUpDownCounter<T>(
+                            string name,
+                            Func<T> observeValue,
+                            string? description = null,
+                            string? unit = null) where T : unmanaged { throw null; }
+
+        public ObservableUpDownCounter<T> CreateObservableUpDownCounter<T>(
+                            string name,
+                            Func<Measurement<T>> observeValue,
+                            string? description = null,
+                            string? unit = null) where T : unmanaged { throw null; }
+
+        public ObservableUpDownCounter<T> CreateObservableUpDownCounter<T>(
+                            string name,
+                            Func<IEnumerable<Measurement<T>>> observeValues,
+                            string? description = null,
+                            string? unit = null) where T : unmanaged { throw null; }
+
+        public void Dispose() { throw null; }
+    }
+}
+```
+
+### Instrument Base Class
+
+```csharp
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// Is the base class which contains all common properties between different types of instruments.
+    /// It contains the protected constructor and the Publish method allows activating the instrument
+    /// to start recording measurements.
+    /// </summary>
+
+    public abstract class Instrument
+    {
+        /// <summary>
+        /// Protected constructor to initialize the common instrument properties.
+        /// </summary>
+        protected Instrument(Meter meter, string name, string? description, string? unit) { throw null; }
+
+        /// <summary>
+        /// Publish is to allow activating the instrument to start recording measurements and to allow
+        /// listeners to start listening to such measurements.
+        /// </summary>
+        protected void Publish() { throw null; }
```

Comment:
Its not clear what is the use case of a protected Publish() method.. when one does meter.CreateCounter/Gauage, won't that be automatically published? 
Do we expect any instrument to not be published?

Reply:
`Publish` is important to tell when start using the instrument after creating and initializing it. We didn't call it automatically inside the abstract base classes constructors because that means the instrument would be published before it is fully initialized in the concrete classes. When using our built-in instruments, you'll not need to call it manually as it will get called from that instrument's constructors. Publish will be needed if someone introduce a custom instrument. They will need to call it from the constructor at that time.

Reply:
Got it. Somehow I misread it as a something end users need to do.
So for the built-in instruments, simply creating them is enough for end user. 

publish() is mostly for anyone trying to create own instruments.